### ISSUE TITLE
Engines check broken for nightly builds

### ISSUE THREAD

**Issue by swernerx** [yarnpkg/yarn#2172]
We have a package which has the following `engines` field:

```
"engines": {
  "node": ">=6.0.0",
  "yarn": ">=0.18.0",
  "npm": ">=4.0.0"
}
```

in `package.json`. As there is not yet a stable "msi" package for Windows users for installation we thought of installing the latest nightly. Unfortunately yarn does not seem to be happy with the installed version. Possibly this is because the version check is somehow flawed.

Output from installing:

```
yarn install v0.18.0-20161123.1451
[1/4] Resolving packages...
[2/4] Fetching packages...
error advanced-boilerplate@0.14.7: The engine "yarn" is incompatible with this module. Expected version ">=0.18.0".
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

This seems to be a bug as `v0.18.0-20161123.1451` should be successfully matched by `>=0.18.0`.

**Comment by Daniel15** [comment]
> As there is not yet a stable "msi" package for Windows users for installation

https://yarnpkg.com/latest.msi will get you the latest stable build.

**Comment by swernerx** [comment]
@Daniel15 thanks, problem is it seems the latest stable differes widely depending on where you look. Homebrew/Git is at 0.18.0 while npm/msi is on 0.17.x.

**Comment by Daniel15** [comment]
0.18.0 is essentially a release candidate.

**Comment by swernerx** [comment]
aha - interesting - It would be probably helpful to follow semver ;)

**Comment by Daniel15** [comment]
Yeah, https://github.com/yarnpkg/yarn/issues/2161 tracks that. There's some
tooling changes that need to be done.

**Comment by gsklee** [comment]
@bestander close

On the original issue that `"yarn": ">=0.18.0"` couldn't match `0.18.0-20161123.1451`, can't do anything here because this is exactly how [`node-semver`](https://github.com/npm/node-semver) works; eg. if you require a `react` version `>=15`, you will not get version `16.0.0-alpha.13`. You should bring this issue to `node-semver` instead.

**Comment by bestander** [comment]
Actually Yarn decided to break semver here after multiple requests from the community.
This should be implemented in 0.25

**Comment by bdwain** [comment]
@bestander was this actually implemented in 0.25? I'm experiencing this issue with the nightly build for 1.4.1 when trying to install https://github.com/sebastian-software/rollup-plugin-rebase (which has

**Comment by bestander** [comment]
It will be hard to find the commit now :(
Try debugging the code and see where the exception is thrown.
Yarn is quite debuggable, the build matches source code quite well

**Comment by bdwain** [comment]
it looks like it still follows semver according to https://github.com/yarnpkg/yarn/blob/879b4f3c893b5e5ec162a1d0a9234c3793e7ac01/src/package-compatibility.js#L59.

> semver.satisfies('1.4.1-20180208.2355', '>=1.0.0', true)
false

Now that the ignore engines flag works, it's not as big of a deal, but if the intent is still to not follow semver for this, I could open a PR.

**Comment by bestander** [comment]
Ah, I see why I am confused.

We chose to ignore this semver feature for dependencies like `react@0.15-beta.2` would satisfy `^0.15` in Yarn but we probably never hacked check for Yarn engine itself.

Please go ahead and send a PR to use our own hacked semver in this case https://github.com/yarnpkg/yarn/blob/879b4f3c893b5e5ec162a1d0a9234c3793e7ac01/src/util/semver.js

**Comment by bdwain** [comment]
Made #5352 for this.
